### ISSUE TITLE
Re-export types and values from other modules

### ISSUE THREAD

**Issue by lpil** [gleam-lang/gleam#614]
It would be nice to be able to re-export a type or function from another module.

Here's an idea for a syntax:

```rust
// one.gleam
pub type MyType {
  MyConstructor
}

pub fn my_function() -> Nil {
  Nil
}
```

```rust
// two.gleam

import two.{pub MyType, pub MyConstructor, pub my_function}
```

## Questions

Should it be permitted to export only a subset of a type's constructors? This would mean it would not possible for users to pattern match on all records as only some have been exported.

**Comment by CrowdHailer** [comment]
I like the syntax, but only rexporting some constructors would be very awkward. (i'm sure I was discussing this somewhere previously but canot find it)

I'm keen that import `option.{Option}` should import Some and None, but I understand that it would mean you couldn't just search in a file to see where the Some and None constructors where defined.

**Comment by lpil** [comment]
Let's use the syntax above as there's no other suggestions or major complaints.

**Comment by lpil** [comment]
> Should it be permitted to export only a subset of a type's constructors? This would mean it would not possible for users to pattern match on all records as only some have been exported. The above syntax would allow that.

This makes me think of @hayleigh-dot-dev's suggestion to use a nested syntax for importing record constructors. I previously dismissed it but there's a second reason to have it here.

**Comment by lpil** [comment]
Adding to v1 as this may require syntax changes

**Comment by pvdrz** [comment]
I'd favor having the `pub` as a qualifier for `import` instead as it would be easier to read which imports are public and which ones are not:
```gleam
import two.{pub MyType, my_function, pub MyConstructor}
```
vs
```gleam
pub import two.{MyType, MyConstructor}
import two.my_function
```

**Comment by inoas** [comment]
I'd also prefer to have separation between pub and non-pub imports as @pvdrz shows.

Reasoning, it may be a bit more verbose but:

- it far far better to read and 
- the formatter could group into two parts, alphabetically sorted pub imports and alphabetically sorted non-pub imports.

**Comment by timjs** [comment]
Fully agree with @pvdrz and @inoas here for reasons they already stated!

**Comment by inoas** [comment]
another question I'd have is: if you re-export a function (or type) but also define it in the same file, what happens?
imho compiler error or newly defined is being exported > reexported is being exported

**Comment by lpil** [comment]
Considering `pub import a.{b}` and `import a.{pub b}` syntaxes.

It seems that `pub import a.{b}` implies you can import `a` from this module in some fashion.

`pub import` means that we have to permit module a module to be imported multiple times. What would the rules be here?

**Comment by massivefermion** [comment]
Sorry to bring this up again but I think it's relevant to your comment.
I've said it before that to my brain, `import a.{b, c}` suggests that I'm only importing `b` and `c`. But gleam imports `a` too.
This is the actual root of the ambiguity you're talking about, not `pub`. Because even with `import a.{pub b}`, `a` is still in scope. So even with `import a.{pub b}`, you should still resolve that ambiguity.
I think given the circumstance, whether we use `pub import a.{b}` or `import a.{pub b}`, we should have `as _` to make it clear to the developer that only `b` is being reexported.
But then it would raise the question, are we reexporting all of `a` if `as _` is not used? It's important to realize that even if we're not doing that, this would be a question that will come to the developer's mind, which is not ideal.
I know introducing new syntax/keywords is frowned upon, but given the mishmash of all the ambiguities here, the best and most clear option is to have a new keyword for this feature:
```
export a.{b, type MyType, MyConstructor}
```
This would mean that if you want to use something in the module but also reexport it, you would need to use both `import` and `export`. This may seem more verbose, but is far more clear as to what is actually going on.
I'm also on the side of allowing the developer to reexport only a subset of constructors.

**Comment by lpil** [comment]
```gleam
import a.{type MyType, type MyType2, MyConstructor, MyConstructor2, func1, func2}
export a.{type MyType, MyConstructor, func1}
```
This seems to imply you can re-export a type without importing it. Is that right?

**Comment by massivefermion** [comment]
Yes, I think that would be more understandable and I don't see any issues. But if you think that may cause issues, you could make it a compile error if something that is not in scope is being exported.

**Comment by lpil** [comment]
To import something it should also be brought into scope as otherwise you'd have a situation in which you can import an item from a module when its not actually in that module's scope.

**Comment by dmmulroy** [comment]
> Considering `pub import a.{b}` and `import a.{pub b}` syntaxes.
> 
> It seems that `pub import a.{b}` implies you can import `a` from this module in some fashion.
> 
> `pub import` means that we have to permit module a module to be imported multiple times, or would the programmer be forced to `pub import a.{b} as _` to not assign a name? What would the rules be here?

What if rather than this syntax and trying to "overload" or extend what "`import`" we introduced something like 

`include { type X, Y } from module`

that would work like an explicit version of OCaml's module inclusion: https://ocaml.org/docs/modules#module-inclusion

This would prevent the confusion of importing a module multiple times or having to rename via `as`. It would also have the benefit of not having to deal with introducing the `pub` keyword to imports.

```gleam
// foo.gleam
pub type Foo {
  Foo(String)
}

pub fn print_foo(foo: Foo) -> Nil { ... }

// bar.gleam
import project/foo.{print_foo}

include { type Foo, Foo } from project/foo

pub type Bar {
  Bar(String)
}


// baz.glem
import project/bar.{type Foo, type Bar, Foo, Bar}
...
```

**Comment by leostera** [comment]
You could probably get away just by swapping the keyword from `import` to `include`, or potentially even a `pub` modifier on the entire `import`:

```gleam
import project/foo.{print_foo}
pub import project/foo.{ type Foo }
```

**Comment by dmmulroy** [comment]
I think the problem with `pub import project/foo.{ type Foo }` is that it bring `foo` into scope. So potentially just swapping `import` with `include` is enough to change the semantics.

`include project/foo.{ type Foo }`

You could use the `pub` keyword to effectively do the same, but I think having a difference between `import` and `include` is nice and might be more clear they work differently

**Comment by iacore** [comment]
Referencing https://github.com/gleam-lang/gleam/issues/2712#issuecomment-1998681052, we would not have have this problem if types are first-class.

**Comment by lpil** [comment]
@iacore If you have a proposal for a language addition please write and share the proposal in GitHub discussions, thank you.

**Comment by iacore** [comment]
Is this invalid or too verbose?

```
import two.{MyType, MyConstructor, my_function}

pub type MyType = two.MyType
pub const MyConstructor = two.MyConstructor
pub const my_function = two.my_function
```

**Comment by lpil** [comment]
Could work! That design seems to imply that one could export a subset of the constructors, or use different names. Those might not be desirable properties.

**Comment by GearsDatapacks** [comment]
It's not really a case of what is possible in the compiler (almost anything is!), but what gives the correct level of control and the desired semantics. 
In this case, the example you provided would be a breaking change. For example, what if you had defined and exported a `Red` value inside `color.gleam`? Which `Red` value would be referenced in a module which imports the `color` module?
Type aliases can also be used today to expose internal types without exposing the constructors. This behaviour would change that, leading to a potential leak of many internal implementation details from packages which exist today.

**Comment by lpil** [comment]
We also aim to always have the names in the code rather than being done implicitly, for ease of understanding.

**Comment by zeddash** [comment]
# `pub import`

The issues I've been facing is that because there's no circular dependencies I can't cleanly have 1 entry point to a group of related modules. I really like writing more but shorter files and Gleam makes it so messy as I'm constantly importing from multiple modules inside a folder.

You're forced to have a module for a type that other modules will use, and whenever you use these modules elsewhere you have to import the type. Or have really long files which is really unenjoyable to deal with, so much scrolling and searching to find the thing I want rather than selecting the module that does the thing.

So, what if re-exporting was less like re-exporting and more like module splitting?

You can ensure there are no circular dependencies by still keeping the flow of imports going downwards so that when modules are merged you don't get a full circle - even though when following imports you can start and end in the same module, but it's the same as `pub type` and `pub const`.

## Example

### foo/types/foo.gleam

```gleam
pub type Foo {
  Foo
  FooBar
  Fooey
}
```

### foo/bar.gleam

```gleam
import foo/types/foo.{type Foo}

@internal
pub fn bar(a: Foo, b: Foo) -> Foo {
  //..
}
```

### foo/foo.gleam

```gleam
// Create the link
pub import foo/types/foo
// ..you will still need to import what you need from the top file so that the file has explicit references
import foo/types/foo.{type Foo}
// `pub import foo/types/foo.{type Foo}` might read as only `type Foo` being re-exported

// even though `foo/types/foo` will be merged with `foo/foo` and `foo/bar` depends on `foo/types/foo` this doesn't create a logical circular dependency
import foo/bar

pub fn foo(a: Foo) -> Foo {
  // ..
}

// could also be a nicer way to re-export functions rather than the `{fun as fun2}` `pub const fun = fun2` method
pub fn bar.bar
pub fn bar.bar as second_bar_same_as_the_first // renaming

```

### then/elsewhere.gleam

```gleam
import foo/foo.{type Foo, Fooey}

pub fn bar_some_foo(a: Foo) -> Foo {
  foo.bar(a, Fooey)
}
```

<img width="800" height="1200" alt="Image" src="https://github.com/user-attachments/assets/d0fe4b91-4616-4990-a3a5-c4724c92d191" />

### Flattening Paths?

It would also be nice if you have a file named the same as the folder it's in that it flattens, `foo/foo` into `foo`. You could have `./foo.gleam` which does all of the re-exporting from `./foo/types/foo.gleam` but it'd be nice to have relevant files together, `import foo.{type Foo, Fooey}` would be really nice.

<sub>I realised too late that I over-used "foo" in this.</sub>

**Comment by lpil** [comment]
How would one export a subset of the module?

**Comment by zeddash** [comment]
You wouldn't, you'd just combine modules. If you don't want to export something, put it somewhere else.

Wouldn't exporting a subset, like only exporting `Fooey` be an issue? Bundling everything marked as `pub` in the exported module would keep important things together.

**Comment by lpil** [comment]
Any system needs to be able to export only a subset to be considered.

It likely also needs to explicitly list all the items to export, to keep with Gleam's rule of having the name of the item in the source any time something is added to the scope.

**Comment by zeddash** [comment]
Fair enough, it's just potentially a lot of repetitive code and having to change re-exporters if you make a change that shouldn't actually affect the module re-exporting it. Not a deal breaker at all tho.

I don't see a solution better than your initial suggestion of

```gleam
import two.{pub MyType, pub MyConstructor, pub my_function}
```

You could move the `pub` to the start. This also has the behaviour of other things made public, that you can expect to not get unused code warnings. However, you may need to import twice for the private stuff which you'd expect to get a duplicate import error (error > warning).

```gleam
pub import two.{MyType, MyConstructor, my_function}
import two.{dont_share_this_function}
```

But I think your initial suggestion is better.

**Comment by lpil** [comment]
APIs in Gleam are always intentional and designed, having the public API of a module change accidentally due to some other module being changed goes against that.

**Comment by isaacharrisholt** [comment]
Personally I'd rather not have something up in the imports, as I find I often glance over them when reading most code, so they're likely to be missed.

We can already re-export functions as constants, but we could potentially be more explicit using the `fn` keyword, which could then indicate to the docs generator that we want to document this as a function (or, to be fair, the behaviour could be changed to do that for `const` functions too).

```gleam
pub fn some_func = internal.some_func
```

Constructors are a bit trickier, but since we're likely always going to want to export all constructors for a type, how about one of these options?

```gleam
pub type ExportedType = internal.InternalType(..)
// or
pub type ExportedType = internal.InternalType(_)
```

Personally I prefer the first, since `..` is used elsewhere in Gleam to mean "all the things," but I'd be fine with either. Could potentially also use `{}` over `()` to better mirror existing type declaration syntax. For some reason though, this feels icky to me. Can't explain why. 

```gleam
pub type ExportedType = internal.InternalType { .. }
```

**Comment by lpil** [comment]
We don't introduce things into scope without using the names, so we would need to be very sure that's the way to go before breaking that rule.
